src/include/homestore/homestore_decl.hpp

```diff
@@ -211,6 +212,36 @@ struct cap_attrs {
 };
 #endif
 
+#pragma pack(1)
+struct chunk_info {
+    static constexpr size_t size = 512;
+    static constexpr size_t user_private_size = 128;
+    static constexpr size_t selector_private_size = 64;
+
+    uint64_t chunk_start_offset{0}; // 0: Start offset of the chunk within a pdev
+    uint64_t chunk_size{0};         // 8: Chunk size
+    uint64_t end_of_chunk_size{0};  // 16: The offset indicates end of chunk.
+    uint32_t vdev_id{0};            // 24: Virtual device id this chunk hosts. UINT32_MAX if chunk is free
+    uint32_t chunk_id{0};           // 28: ID for this chunk - unique for entire homestore across devices
+    uint32_t chunk_ordinal{0};      // 32: Chunk ordinal within the vdev on this pdev
+    uint8_t chunk_allocated{0x00};  // 36: Is chunk allocated or free
+    uint16_t checksum{0};           // 37: checksum of this chunk info
+    uint8_t padding[25]{};          // 39: pad to make it 128 bytes total
+    uint8_t chunk_selector_private[selector_private_size]{}; // 64: Chunk selector private area
+    uint8_t user_private[user_private_size]{};               // 128: Opaque user of the chunk information
+
+    uint64_t get_chunk_size() const;
+    uint32_t get_chunk_id() const;
+    bool is_allocated() const;
+    void set_allocated();
+    void set_free();
+
+    void set_selector_private(const sisl::blob& data);
+    void set_user_private(const sisl::blob& data);
+    void compute_checksum();
+};
+#pragma pack()
+
```

Comment:
Again, no layer above HomeStore needs to see these fields and structure (I hope); is this all needed by the HeapChunkSelector? If not can we prune it similarly or remove it entirely from the API?

Reply:
yea, i do not want is here too. as I mentioned above , i have to expose "class Chunk", and  it has "chunk_info m_chunk_info", so it is necessary to sizeof(chunk_info) and we have to expose it.

a solution is to use `shared_ptr<chunk_info>` instead of `chunk_info`, and this need more modification in homestore to change all the reference to this. if you agree with this idea, i will do this change

Reply:
We should certainly not put in homestore_decl.hpp. I don't think we need to change `chunk_info` to `std::shared_ptr< chunk_info >`, if you can define Chunk itself as a abstract class (say AbstractChunk) and then Chunk could be derived and can have chunk_info() native to Chunk and inside device/ folder

Reply:
Yes then the `ChunkSelector` can only return `AbstractChunk`,  the caller need to static cast that to `Chunk` as it is from base to derived class

Reply:
```
class PhysicalDev;
class baseChunk {
public:
    baseChunk() = defalut;
    virtual ~baseChunk() = default;

    virtual void set_user_private(const sisl::blob& data);
    virtual uint8_t* get_user_private();
    virtual blk_cap_t available_blks();
    virtual PhysicalDev* getPhysicalDev();
}
```

```
class ChunkSelector {
public:
    ChunkSelector() = default;
    virtual void add_chunk(cshared< baseChunk >& chunk) = 0;
    virtual void foreach_chunks(std::function< void(cshared< baseChunk >&) >&& cb) = 0;
    virtual baseChunk* select(blk_count_t nblks, const blk_alloc_hints& hints) = 0;

    virtual ~ChunkSelector() = default;
};
```

i have considered exposing the two classes in the header file earlier. the only problem is that , as @xiaoxichen said , we have to do static_cast from a base class to derived class when calling `select` in vdev, which seems not graceful.

but seems this will make the exposed header more simple. 

@hkadayam @xiaoxichen are you ok to expose the two classes in the header file , and do the static_cast inside vdev after calling `chunkselector#selector`